gapic/samplegen/samplegen.py

```diff
@@ -0,0 +1,376 @@
+    pass
+
+
+class InvalidRequestSetup(SampleError):
+    pass
+
+
+class DuplicateInputParameter(SampleError):
+    pass
+
+
+class ResponseValidator:
+    COLL_KWORD = "collection"
+    VAR_KWORD = "variable"
+    MAP_KWORD = "map"
+    KEY_KWORD = "key"
+    VAL_KWORD = "value"
+    BODY_KWORD = "body"
+
+    def __init__(self, response):
+        """Validates a "response" block from a sample config.
+
+        A full description of the response block is outside the scope of this code
+        and is better handled by reading the samplegen documentation.
+
+        Args:
+           response: list[dict{str:?}]: The structured data representing
+                                        the sample's response.
+
+        Raises:
+            BadAssignment: If a "define" statement is badly formed.
+            ReservedVariablename: If an attempted lvalue is a reserved word.
+            UndefinedVariableReference: If an attempted rvalue base is a previously
+                                        undeclared variable.
+            MismatchedPrintStatement: If the number of format string segments ("%s") in
+                                      a "print" or "comment" block does not equal the
+                                      size number of strings in the block minus 1.
+            BadLoop: If a "loop" segments has unexpected keywords
+                     or keyword combinatations.
+            InvalidStatement: If an unexpected key is found in a statement dict
+                              or a statement dict has more than or less than one key.
+        """
+
+        # The response ($resp) variable is special and guaranteed to exist.
+        self.var_defs_ = {"$resp"}
+
+        self.inner_validate_(response)
+
+    def inner_validate_(self, response):
+        for statement in response:
+            if len(statement) != 1:
+                raise InvalidStatement(
+                    "Invalid statement: {}".format(statement))
+
+            keyword, body = next(iter(statement.items()))
+            validater = self.STATEMENT_DISPATCH_TABLE.get(keyword)
+            if not validater:
+                raise InvalidStatement("Invalid statement keyword: {}"
+                                       .format(keyword))
+
+            validater(self, body)
+
+    def validate_comment_(self, body: List[str]):
+        fmt_str = body[0]
+        num_vars = fmt_str.count("%s")
+        if num_vars != len(body) - 1:
+            raise MismatchedPrintStatement(
+                "Expected {} var names in comment but received {}"
+                .format(num_vars, len(body) - 1))
+
+        for var_name in body[1:]:
+            var_base = var_name.split(".")[0]
+            if var_base not in self.var_defs_:
+                raise UndefinedVariableReference("Reference to undefined variable: {}"
+                                                 .format(var_base))
+
+    def validate_print_(self, body: List[str]):
```

Comment:
Aren't `validate_print` and `validate_comment_` doing the same thing? I would just have one such validator and dispatch to it for both `print` and `comment`

Reply:
Looks like it. Artifact of TDD.